### ISSUE TITLE
Error occurred when testing example1 with device hsa 

### ISSUE THREAD

**Issue by 0charleschen0** [pocl/pocl#190]
I tested example1 with device hsa and I got this error:
 **error: unsupported initializer for address space in dot_product**
in pocl_llvm_codegen() in compile) in pocl_hsa_compile_submitted_kernels().

This is program.bc disassembled from program.ll
```Assembly
; ModuleID = 'program.bc'
target datalayout = "e-p:32:32-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-p24:64:64-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
target triple = "amdgcn--amdhsa"

; Function Attrs: nounwind
define void @dot_product(<4 x float> addrspace(1)* nocapture readonly %a, <4 x float> addrspace(1)* nocapture readonly %b, float addrspace(1)* nocapture %c) #0 {
entry:
  %call = tail call i32 @_Z13get_global_idj(i32 0) #2
  %0 = sext i32 %call to i64
  %arrayidx = getelementptr inbounds <4 x float> addrspace(1)* %a, i64 %0
  %1 = load <4 x float> addrspace(1)* %arrayidx, align 16, !tbaa !8
  %arrayidx1 = getelementptr inbounds <4 x float> addrspace(1)* %b, i64 %0
  %2 = load <4 x float> addrspace(1)* %arrayidx1, align 16, !tbaa !8
  %mul = fmul <4 x float> %1, %2
  %3 = extractelement <4 x float> %mul, i32 0
  %4 = extractelement <4 x float> %mul, i32 1
  %add = fadd float %3, %4
  %5 = extractelement <4 x float> %mul, i32 2
  %add2 = fadd float %5, %add
  %6 = extractelement <4 x float> %mul, i32 3
  %add3 = fadd float %6, %add2
  %arrayidx4 = getelementptr inbounds float addrspace(1)* %c, i64 %0
  store float %add3, float addrspace(1)* %arrayidx4, align 4, !tbaa !11
  ret void
}

declare i32 @_Z13get_global_idj(i32) #1

attributes #0 = { nounwind "less-precise-fpmad"="false" "no-frame-pointer-elim"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #1 = { "less-precise-fpmad"="false" "no-frame-pointer-elim"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #2 = { nobuiltin nounwind }

!opencl.kernels = !{!0}
!llvm.ident = !{!7}

!0 = !{void (<4 x float> addrspace(1)*, <4 x float> addrspace(1)*, float addrspace(1)*)* @dot_product, !1, !2, !3, !4, !5, !6}
!1 = !{!"kernel_arg_addr_space", i32 1, i32 1, i32 1}
!2 = !{!"kernel_arg_access_qual", !"none", !"none", !"none"}
!3 = !{!"kernel_arg_type", !"float4*", !"float4*", !"float*"}
!4 = !{!"kernel_arg_base_type", !"float __attribute__((ext_vector_type(4)))*", !"float __attribute__((ext_vector_type(4)))*", !"float*"}
!5 = !{!"kernel_arg_type_qual", !"const", !"const", !""}
!6 = !{!"kernel_arg_name", !"a", !"b", !"c"}
!7 = !{!"clang version 3.6.0 (tags/RELEASE_360/final)"}
!8 = !{!9, !9, i64 0}
!9 = !{!"omnipotent char", !10, i64 0}
!10 = !{!"Simple C/C++ TBAA"}
!11 = !{!12, !12, i64 0}
!12 = !{!"float", !9, i64 0}
```
and this is parallel.ll disassembled from parallel.bc
```Assembly
; ModuleID = 'parallel.bc'
target datalayout = "e-p:32:32-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-p24:64:64-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
target triple = "amdgcn--amdhsa"

@_global_offset_x = external global i32
@_local_size_x = external global i32
@_group_id_x = external global i32

; Function Attrs: nounwind readnone
declare i32 @llvm.r600.read.tidig.x() #0

; Function Attrs: noinline nounwind
define void @dot_product(<4 x float> addrspace(1)* nocapture readonly %a, <4 x float> addrspace(1)* nocapture readonly %b, float addrspace(1)* nocapture %c) #1 {
entry:
  %0 = load i32* @_global_offset_x, align 4
  %1 = load i32* @_local_size_x, align 4
  %2 = load i32* @_group_id_x, align 4
  %mul.i = mul i32 %2, %1
  %add.i = add i32 %mul.i, %0
  %x.i.i = tail call i32 @llvm.r600.read.tidig.x() #0
  %add1.i = add i32 %add.i, %x.i.i
  %3 = sext i32 %add1.i to i64
  %arrayidx = getelementptr inbounds <4 x float> addrspace(1)* %a, i64 %3
  %4 = load <4 x float> addrspace(1)* %arrayidx, align 16, !tbaa !2
  %arrayidx1 = getelementptr inbounds <4 x float> addrspace(1)* %b, i64 %3
  %5 = load <4 x float> addrspace(1)* %arrayidx1, align 16, !tbaa !2
  %mul = fmul <4 x float> %4, %5
  %6 = extractelement <4 x float> %mul, i32 0
  %7 = extractelement <4 x float> %mul, i32 1
  %add = fadd float %6, %7
  %8 = extractelement <4 x float> %mul, i32 2
  %add2 = fadd float %8, %add
  %9 = extractelement <4 x float> %mul, i32 3
  %add3 = fadd float %9, %add2
  %arrayidx4 = getelementptr inbounds float addrspace(1)* %c, i64 %3
  store float %add3, float addrspace(1)* %arrayidx4, align 4, !tbaa !5
  ret void
}

attributes #0 = { nounwind readnone }
attributes #1 = { noinline nounwind "less-precise-fpmad"="false" "no-frame-pointer-elim"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }

!llvm.ident = !{!0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0, !0}
!opencl.kernels = !{!1}

!0 = !{!"clang version 3.6.0 (tags/RELEASE_360/final)"}
!1 = !{void (<4 x float> addrspace(1)*, <4 x float> addrspace(1)*, float addrspace(1)*)* @dot_product}
!2 = !{!3, !3, i64 0}
!3 = !{!"omnipotent char", !4, i64 0}
!4 = !{!"Simple C/C++ TBAA"}
!5 = !{!6, !6, i64 0}
!6 = !{!"float", !3, i64 0}
```
for reference about the error.

**Comment by pjaaskel** [comment]
Did you use LLVM 3.6 and the "legacy" HSA runtime?

**Comment by 0charleschen0** [comment]
Yes, it is.

**Comment by pjaaskel** [comment]
These cause the issue for you:
```
@_global_offset_x = external global i32
@_local_size_x = external global i32
@_group_id_x = external global i32
```

For some reason the correct versions of the HSA kernel lib are not linked in for you. It works for me (just tested it now) and I get a parallel.ll like this for HSA:
```
; ModuleID = 'tmp/MG/KEHCGFCNOFMNKEDJLLIDKNDODEOEFEACCCHBA/dot_product/128-1-1/parallel.bc'
target datalayout = "e-p:32:32-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-p24:64:64-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
target triple = "amdgcn--amdhsa"

; Function Attrs: nounwind readnone
declare i32 @llvm.r600.read.local.size.x() #0

; Function Attrs: nounwind readnone
declare i32 @llvm.r600.read.tgid.x() #0

; Function Attrs: nounwind readnone
declare i32 @llvm.r600.read.tidig.x() #0

; Function Attrs: noinline nounwind
define void @dot_product(<4 x float> addrspace(1)* nocapture readonly %a, <4 x float> addrspace(1)* nocapture readonly %b, float addrspace(1)* nocapture %c) #1 {
entry:
  %x.i.i = tail call i32 @llvm.r600.read.local.size.x() #0
  %x.i6.i = tail call i32 @llvm.r600.read.tgid.x() #0
  %mul.i = mul i32 %x.i6.i, %x.i.i
  %x.i5.i = tail call i32 @llvm.r600.read.tidig.x() #0
  %add.i = add i32 %mul.i, %x.i5.i
  %0 = sext i32 %add.i to i64
  %arrayidx = getelementptr inbounds <4 x float> addrspace(1)* %a, i64 %0
  %1 = load <4 x float> addrspace(1)* %arrayidx, align 16, !tbaa !2
  %arrayidx1 = getelementptr inbounds <4 x float> addrspace(1)* %b, i64 %0
  %2 = load <4 x float> addrspace(1)* %arrayidx1, align 16, !tbaa !2
  %mul = fmul <4 x float> %1, %2
  %3 = extractelement <4 x float> %mul, i32 0
  %4 = extractelement <4 x float> %mul, i32 1
  %add = fadd float %3, %4
  %5 = extractelement <4 x float> %mul, i32 2
  %add2 = fadd float %5, %add
  %6 = extractelement <4 x float> %mul, i32 3
  %add3 = fadd float %6, %add2
  %arrayidx4 = getelementptr inbounds float addrspace(1)* %c, i64 %0
  store float %add3, float addrspace(1)* %arrayidx4, align 4, !tbaa !5
  ret void
}
...
```

Is your lib/kernel/amdgcn built correctly?

**Comment by pjaaskel** [comment]
I have somewhat newer LLVM, but it should not be the cause here. It's more likely that your builtin lib is not correctly built.  

```
!0 = !{!"clang version 3.6.1 (branches/release_36 233533)"}
```

**Comment by 0charleschen0** [comment]
```
/home/moea-hsa/llvm-3.6/build/bin/clang -Xclang -ffake-address-space-map -emit-llvm -target amdgcn--amdhsa -mcpu=kaveri -D__OPENCL_VERSION__=120 -D__CBUILD__ -c -o get_global_id.c.bc -include /home/moea-hsa/pocl/build-pocl-make/../pocl/include/_kernel_c.h ../../../../pocl/lib/kernel/get_global_id.c
```
This is the command to generate get_global_id.c.bc which causes the issue.
Is it the correct build command?

**Comment by 0charleschen0** [comment]
It seems that *get_global_id.c.bc* built from **kernel/**get_global_id.c instead of **kernel/amdgcn/**get_global_id.c
How should I fix it?

**Comment by pjaaskel** [comment]
Are you using Cmake or automake? This might be an automake version issue. automake 1.11 is known to work

**Comment by 0charleschen0** [comment]
I'm using automake 1.14.1. Should I downgrade to 1.11?
About CMake, I think CMake should be modified to build hsa kernel library.

**Comment by pjaaskel** [comment]
I have automake 1.14.1 also -- this is strange. ./autogen.sh worked for you? You are right CMake is not yet updated for the HSA.

**Comment by pjaaskel** [comment]
make get_global_id.c.bc
mkdir -p ./
/home/visit0r/local/stow/llvm-3.6/bin/clang -Xclang -ffake-address-space-map -emit-llvm -target amdgcn--amdhsa -mcpu=kaveri -D__OPENCL_VERSION__=120 -D__CBUILD__ -c -o get_global_id.c.bc -include /home/visit0r/src/pocl-public/include/_kernel_c.h get_global_id.c

**Comment by 0charleschen0** [comment]
./autogen.sh worked for me.
Below is the information printed out from autogen.sh for my environment.
```
Found GNU Autoconf version 2.69
Found GNU Automake version 1.14.1
Found GNU Libtool version 2.4.2
```

**Comment by pjaaskel** [comment]
OK, you are using out-of-source-tree build which I don't usually use. Testing it...

**Comment by pjaaskel** [comment]
Right, that was it. Fixed in ab052d2.
